### PR TITLE
Add `build_callgraph` to `UnaryIterationGate`

### PR DESCRIPTION
**Pull request by tanujkhattar**
Fixes https://github.com/quantumlib/Qualtran/issues/559
Inspired from https://github.com/quantumlib/Qualtran/pull/609

### CODE REVIEW

File: qualtran/bloqs/unary_iteration_bloq.py [hunk 1]
```diff
@@ -275,34 +275,69 @@ def _unary_iteration_callgraph_segtree(
...
+            subtree of the node representing range `[l, r)`. If True, the internal node is
+            considered equivalent to a leaf node and the method yields only one tuple
+            `(OP_TREE, control_qubit, l)` for all integers in the range `[l, r)`.
+        bloq_counts: Mutable Dictionary to which the counts of bloqs used by the unary iteration
+            segment tree are appended.
+
+    Returns:
+        Returns a sequence of integers, each representing the first element `l` in the range
+        `[l, r)` corresponding to the leaf nodes of the constructed segment tree.
+
```
Comment (fdmalone) [inline qualtran/bloqs/unary_iteration_bloq.py:304]: nit: remove new line.

File: qualtran/bloqs/unary_iteration_bloq.py [hunk 2]
```diff
@@ -275,34 +275,69 @@ def _unary_iteration_callgraph_segtree(
...
+            # We don't need to traverse this subtree since it's parent was already "marked".
             continue
         if l[i] >= r_iter or l_iter >= r[i]:
             # Range corresponding to this node is completely outside of iteration range.
             marked[i] = 1
             continue
-        if l_iter <= l[i] < r[i] <= r_iter and (l[i] == (r[i] - 1) or break_early(l[i], r[i])):
-            # Reached a leaf node or a "special" internal node; yield the operations.
+        if l_iter <= l[i] < r[i] <= r_iter and (i >= n or break_early(l[i], r[i])):
+            # Reached a leaf node or a "special" internal node; append it's left element.
```
Comment (fdmalone) [inline qualtran/bloqs/unary_iteration_bloq.py:331]: nit, its left element.

**Review by fdmalone — APPROVED** [review]
Huge improvement in runtime. Thanks!

We could just write down the number of And's and X's BUT this is complicated by the variable spaced qrom implementation, which is probably hard to work out in general.

**Comment by tanujkhattar** [comment]
ah, I just pushed another improvement. I think it got left out. I'll open a PR

**Pull request merged**